_posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

```diff
@@ -0,0 +1,757 @@
+
+With just a single layer, `layer_outputs[[1]]`holds all of the layer's intermediate outputs, stacked on dimension two.
+
+```{r}
+dim(layer_outputs[[1]])
+# [1]  2  4  5 16 16
+```
+
+`layer_last_states[[1]]`is a list of tensors, the first of which holds the single layer's final hidden state, and the second, its final cell state.
+
+```{r}
+dim(layer_last_states[[1]][[1]])
+# [1]  2  5 16 16
+dim(layer_last_states[[1]][[2]])
+# [1]  2  5 16 16
+```
+
+For comparison, this is how return values look for a multi-layer architecture.
+
+```{r}
+model <- convlstm(input_dim = 3, hidden_dims = c(5, 5, 1), kernel_sizes = rep(3, 3), n_layers = 3)
+c(layer_outputs, layer_last_states) %<-% model(x)
+
+# for each layer, tensor of size (batch_size, seq_len, hidden_size, height, width)
+dim(layer_outputs[[1]])
+# 2  4  5 16 16
+dim(layer_outputs[[3]])
+# 2  4  1 16 16
+
+# list of 2 tensors for each layer
+length(layer_last_states)
+# 3
+
+# h, of size (batch_size, hidden_size, height, width)
+dim(layer_last_states[[3]][[1]])
+# 2  1 16 16
+
+# c, of size (batch_size, hidden_size, height, width)
+dim(layer_last_states[[3]][[2]])
+# 2  1 16 16
+```
+
+Now we want to sanity-check this module with the simplest-possible dummy data.
+
+### Sanity-checking the `convlstm`
+
+We generate black-and-white "movies" of diagonal beams successively translated in space.
+
+Each sequence consists of six time steps, and each beam of six pixels. Just a single sequence is created manually. To create that one sequence, we start from a single beam:
+
+```{r}
+library(torchvision)
+
+beams <- vector(mode = "list", length = 6)
+beam <- torch_eye(6) %>% nnf_pad(c(6, 12, 12, 6)) # left, right, top, bottom
+beams[[1]] <- beam
+```
+
+Using `torch_roll()` , we create a pattern where this beam moves up diagonally, and stack the individual tensors along the `timesteps` dimension.
+
+```{r}
+for (i in 2:6) {
+  beams[[i]] <- torch_roll(beam, c(-(i-1),i-1), c(1, 2))
+}
+
+init_sequence <- torch_stack(beams, dim = 1)
+```
+
+That's a single sequence. Thanks to `torchvision::transform_random_affine()`, we almost effortlessly produce a dataset of a hundred sequences. Moving beams start at random points in the spatial frame, but they all share that upward-diagonal motion.
+
+```{r}
+sequences <- vector(mode = "list", length = 100)
+sequences[[1]] <- init_sequence
+
+for (i in 2:100) {
+  sequences[[i]] <- transform_random_affine(init_sequence, degrees = 0, translate = c(0.5, 0.5))
+}
+
+input <- torch_stack(sequences, dim = 1)
+dim(input)
+# [1] 100   6  24  24
+```
+
+That's it for the raw data. Now we still need a `dataset` and a `dataloader`. Of the six time steps, we use the first five as input and try to predict the last one.
+
+```{r}
+dummy_ds <- dataset(
+  
+  initialize = function(data) {
+    # add channels dimension
+    self$data <- data$unsqueeze(3)
```

Comment:
i found it confusing for this to be there. I would make the `input` sequence with the correct size isntead.

Reply:
makes sense!